RN: Implement `sendAccessibilityEvent` in RN Renderer that proxies between Fabric/non-Fabric

JoshuaGross opened the pull request:
Implement `sendAccessibilityEvent` on HostComponent for Fabric and non-Fabric RN.

Currently the Fabric version is a noop and non-Fabric uses
AccessibilityInfo directly. The Fabric version will be updated once
native Fabric Android/iOS support this method in the native UIManager.

<!--
  Thanks for submitting a pull request!
  We appreciate you spending the time to work on these changes. Please provide enough information so that others can review your pull request. The three fields below are mandatory.

  Before submitting a pull request, please make sure the following is done:

  1. Fork [the repository](https://github.com/facebook/react) and create your branch from `master`.
  2. Run `yarn` in the repository root.
  3. If you've fixed a bug or added code that should be tested, add tests!
  4. Ensure the test suite passes (`yarn test`). Tip: `yarn test --watch TestName` is helpful in development.
  5. Run `yarn test --prod` to test in the production environment. It supports the same options as `yarn test`.
  6. If you need a debugger, run `yarn debug-test --watch TestName`, open `chrome://inspect`, and press "Inspect".
  7. Format your code with [prettier](https://github.com/prettier/prettier) (`yarn prettier`).
  8. Make sure your code lints (`yarn lint`). Tip: `yarn linc` to only check changed files.
  9. Run the [Flow](https://flowtype.org/) type checks (`yarn flow`).
  10. If you haven't already, complete the CLA.

  Learn more about contributing: https://reactjs.org/docs/how-to-contribute.html
-->

## Summary

We are deprecating direct calls to RN UIManager methods in favor of methods on HostComponents. 

## Test Plan

Tested in React Native directly - more details tbd

TheSavior commented:
I'd expect this to fail Flow without the changes to ReactNativeTypes. Also, you'll want to add some unit tests for these (React has unit tests!) You can check out the history of those files to find previous PRs I've made for examples of how to write tests

TheSavior commented on packages/react-native-renderer/src/ReactNativeFiberHostComponent.js:
```diff
@@ -124,6 +125,14 @@ class ReactNativeFiberHostComponent {
+  sendAccessibilityEvent(eventType: string) {
+    if (eventType === 'focus') {
+      DeprecatedAccessibilityInfo.setAccessibilityFocus(this._nativeTag);
+    } else {
```
If Fabric is not going to throw on invalid strings from JS, we probably want that behavior consistent on paper. Probably just no-op. Enforcement can come from the types

TheSavior commented:
If you'd like to roll this out incrementally, you could add the Fabric implementation behind a React feature flag, and if the flag is false have it use the Paper implementation. Then you can migrate JS callsites to the new api now without waiting for the native fabric code to have that method yet.

sebmarkbage commented:
Shouldn’t we be deprecating method calls on host instances in favor of module level calls that pass an instance anyway?

Why do this in two steps instead of just deprecating the calls directly in both places and add the third? You’d make people migrate to this only to migrate away.

kacieb commented:
> Shouldn’t we be deprecating method calls on host instances in favor of module level calls that pass an instance anyway?
> 
> Why do this in two steps instead of just deprecating the calls directly in both places and add the third? You’d make people migrate to this only to migrate away later.

Just trying to understand this - does that mean we no longer want to be calling functions like `ref.measure`, `ref.blur()` etc? Are we planning to deprecate these too? What's the new way?

sebmarkbage commented:
I'm not sure what the latest API proposal is but it should be something like instead of `ref.measure(...)` it would be `LayoutSomethingSomething.measure(ref, ...)`, and instead of `ref.blur()` it would be `FocusManagerOrSomething.blur(ref)`.

JoshuaGross commented:
@sebmarkbage @kacieb

Eli's proposal is here: https://github.com/react-native-community/discussions-and-proposals/issues/72

I will be updating this PR to remove methods from the HostComponent API, and follow a pattern similar to viewCommands.

TheSavior commented:
Heh, I actually kinda forgot about that proposal @sebmarkbage, thanks for the reminder. I agree that we probably shouldn't add new methods like that if we can avoid it. 

@JoshuaGross, unit tests will be pretty important for this. You can check out this commit of mine to see how to write them. https://github.com/facebook/react/commit/8533c0a168138d2442334021751b42e530d08bb8#diff-5c608331d700d268ea959ed0d77e87844798b8f8e30222f8ffaa990cac3603d2

Since this code lives on the renderer there are a couple things the tests are checking for:
Passing a ref to a fabric component to the Fabric JS renderer calls the fabric native method
Passing a ref to a paper component to the Paper JS renderer calls the paper native method
Passing a ref to a paper component to the Fabric JS renderer calls the paper native method (these are the edge cases you may not have thought of)
Passing a ref to a fabric component to the Paper JS renderer calls the fabric native method (these are the edge cases you may not have thought of)

JoshuaGross commented:
Made some updates. Will add tests later.

JoshuaGross commented:
fixed flow, ran yarn prettier, added unit tests as @TheSavior suggested: 

* Passing a ref to a fabric component to the Fabric JS renderer calls the fabric native method
* Passing a ref to a paper component to the Paper JS renderer calls the paper native method
* Passing a ref to a paper component to the Fabric JS renderer calls the paper native method
* Passing a ref to a fabric component to the Paper JS renderer calls the fabric native method
* Passing a non-ref to Paper JS render is an error
* Passing a non-ref to Fabric JS render is an error
